File: tests/fixtures/jsts/crossfn_vuln.ts
// Vulnerable fixture for cross-function taint (PATH_TRAVERSAL, feature 012, JS/TS).
// The path-like param flows into a named helper whose param is NOT path-like,
// and the helper reads it without validation. Named function (not arrow const)
// so the port can resolve the callee.
import * as fs from "fs";

export function readFile(filename: string): string {
  return load(filename);
}

function load(x: string): string {
  return fs.readFileSync(x, "utf-8"); // sink reached via cross-function taint
}

File: tests/fixtures/crossfn_vuln.py
"""Vulnerable fixture for cross-function taint (PATH_TRAVERSAL, feature 012).

The path-like tool parameter flows into a helper whose parameter is NOT
path-like-named, and the helper opens it without validation. The old
intra-procedural logic misses this (source and sink are in different
functions); the cross-function pass must catch both the positional and the
keyword flow.
"""


def read_file(filename):
    # positional: filename (path-like source) -> helper _load(x) -> open(x)
    return _load(filename)


def _load(x):
    with open(x) as f:  # sink reached via cross-function taint
        return f.read()


def fetch_doc(filepath):
    # keyword: filepath -> helper _fetch(target=...) -> open(target)
    return _fetch(target=filepath)


def _fetch(target):
    with open(target) as f:  # sink reached via keyword-bound cross-function taint
        return f.read()
